controlplane/kubeadm/internal/controllers/controller.go

```diff
@@ -578,6 +583,71 @@ func (r *KubeadmControlPlaneReconciler) reconcileEtcdMembers(ctx context.Context
 	return ctrl.Result{}, nil
 }
 
+func (r *KubeadmControlPlaneReconciler) reconcileCertificateExpiries(ctx context.Context, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Return if there are no KCP-owned control-plane machines.
+	if controlPlane.Machines.Len() == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// Ignore machines which are being deleted.
+	machines := controlPlane.Machines.Filter(collections.Not(collections.HasDeletionTimestamp))
+
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(controlPlane.Cluster))
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile certificate expiries: cannot get remote client to workload cluster")
+	}
+
+	for _, m := range machines {
+		log = log.WithValues("Machine", klog.KObj(m))
+
+		kubeadmConfig, ok := controlPlane.GetKubeadmConfig(m.Name)
+		if !ok {
+			// Skip if the Machine doesn't have a KubeadmConfig.
+			continue
+		}
+
+		annotations := kubeadmConfig.GetAnnotations()
+		if _, ok := annotations[clusterv1.MachineCertificatesExpiryDateAnnotation]; ok {
+			// Skip if annotation is already set.
+			continue
+		}
+
+		if m.Status.NodeRef == nil {
+			// Skip if the Machine is still provisioning.
+			continue
+		}
+		nodeName := m.Status.NodeRef.Name
+		log = log.WithValues("Node", klog.KRef("", nodeName))
```

Comment:
Should we also add the machine name if it doesn't have it already?

Reply:
Should be covered by l.603
> 		log = log.WithValues("Machine", klog.KObj(m))